core/src/main/scala/kafka/log/ProducerStateManager.scala

```diff
@@ -753,6 +753,18 @@ class ProducerStateManager(val topicPartition: TopicPartition,
       None
   }
 
-  private def listSnapshotFiles: Seq[File] = ProducerStateManager.listSnapshotFiles(logDir)
+  private[log] def listSnapshotFiles: Seq[File] = ProducerStateManager.listSnapshotFiles(logDir)
 
+  /**
+   * Remove any producer state snapshot files which do not have a corresponding offset provided
+   * in keepOffsets. The latest snapshot file will always be kept.
+   */
+  def cleanupOrphanSnapshotFiles(segmentBaseOffsets: Iterable[Long]): Unit = {
+    val expected = segmentBaseOffsets.map(Log.producerSnapshotFile(logDir, _)).toSet ++ latestSnapshotFile
```

Comment:
Just to be clear. You mentioned 

"I choose to implement #2 as it was simpler, and instead of just deleting the highest producer state snapshot file without a corresponding segment file, I delete all producer state snapshot files without a corresponding segment file."

But, we are still keeping the snapshot file with the largest offset?

Reply:
Yes this is a mistake, now that `cleanupOrphanSnapshotFiles` occurs after loading the producer state, we should also delete the latest snapshot file.